File: core/project.py
"""
Project state management — tracks the configuration and progress
of an AERMET automation run.
"""

import json
import os
import time
import uuid

_PROJECTS_DIR = os.path.join(os.path.dirname(__file__), '..', 'projects')


def create_project(location, data_period, completeness, is_us=True):
    """
    Create a new project workspace.

    Args:
        location: dict with lat, lon, display_name
        data_period: dict with num_years, start_year
        completeness: dict with threshold, basis ('quarterly' or 'annual')
        is_us: whether the site is in the US

    Returns project dict with id and paths.
    """
    project_id = str(uuid.uuid4())[:8]
    project_dir = os.path.join(_PROJECTS_DIR, project_id)
    os.makedirs(project_dir, exist_ok=True)
    os.makedirs(os.path.join(project_dir, 'surface'), exist_ok=True)
    os.makedirs(os.path.join(project_dir, 'upper_air'), exist_ok=True)
    os.makedirs(os.path.join(project_dir, 'one_minute'), exist_ok=True)
    os.makedirs(os.path.join(project_dir, 'aermet'), exist_ok=True)
    os.makedirs(os.path.join(project_dir, 'output'), exist_ok=True)

    start_year = data_period['start_year']
    num_years = data_period['num_years']

    project = {
        'id': project_id,
        'dir': project_dir,
        'created': time.time(),
        'location': location,
        'data_period': {
            'start_year': start_year,
            'end_year': start_year + num_years - 1,
            'num_years': num_years,
        },
        'completeness': completeness,
        'is_us': is_us,
        'surface_station': None,
        'upper_air_station': None,
        'status': 'created',
        'steps_completed': [],
        'errors': [],
        'results': {},
    }

    _save_project(project)
    return project


def load_project(project_id):
    """Load project state from disk."""
    path = os.path.join(_PROJECTS_DIR, project_id, 'project.json')
    if not os.path.exists(path):
        raise ValueError(f'Project {project_id} not found')
    with open(path, 'r') as f:
        return json.load(f)


def update_project(project, **kwargs):
    """Update project fields and save."""
    project.update(kwargs)
    _save_project(project)
    return project


def _save_project(project):
    """Persist project state to JSON."""
    path = os.path.join(project['dir'], 'project.json')
    with open(path, 'w') as f:
        json.dump(project, f, indent=2)

File: aermet-tool/src/project.rs
/// Project state management — tracks configuration and progress of an AERMET run.
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};
use uuid::Uuid;

/// Base directory for all projects (resolved at runtime relative to exe).
fn projects_dir() -> PathBuf {
    let exe_dir = std::env::current_exe()
        .ok()
        .and_then(|p| p.parent().map(|p| p.to_path_buf()))
        .unwrap_or_else(|| PathBuf::from("."));
    exe_dir.join("projects")
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Location {
    pub lat: f64,
    pub lon: f64,
    #[serde(default)]
    pub display_name: String,
    #[serde(default)]
    pub country_code: String,
    #[serde(default)]
    pub is_us: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataPeriod {
    pub start_year: i32,
    pub end_year: i32,
    pub num_years: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Completeness {
    pub threshold: f64,
    pub basis: String, // "quarterly" or "annual"
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Project {
    pub id: String,
    pub dir: String,
    pub created: f64,
    pub location: Location,
    pub data_period: DataPeriod,
    pub completeness: Completeness,
    pub is_us: bool,
    pub surface_station: Option<serde_json::Value>,
    pub upper_air_station: Option<serde_json::Value>,
    pub status: String,
    #[serde(default)]
    pub steps_completed: Vec<String>,
    #[serde(default)]
    pub errors: Vec<String>,
    #[serde(default)]
    pub results: serde_json::Value,
}

#[derive(Debug, Deserialize)]
pub struct CreateProjectRequest {
    pub location: Location,
    pub data_period: CreateDataPeriod,
    pub completeness: Completeness,
    #[serde(default = "default_true")]
    pub is_us: bool,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Deserialize)]
pub struct CreateDataPeriod {
    pub start_year: i32,
    pub num_years: i32,
}

pub fn create_project(req: CreateProjectRequest) -> Result<Project, String> {
    let project_id = Uuid::new_v4().to_string()[..8].to_string();
    let project_dir = projects_dir().join(&project_id);

    for sub in &["surface", "upper_air", "one_minute", "aermet", "output"] {
        fs::create_dir_all(project_dir.join(sub))
            .map_err(|e| format!("Failed to create directory: {e}"))?;
    }

    let end_year = req.data_period.start_year + req.data_period.num_years - 1;
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();

    let project = Project {
        id: project_id,
        dir: project_dir.to_string_lossy().to_string(),
        created: now,
        location: req.location,
        data_period: DataPeriod {
            start_year: req.data_period.start_year,
            end_year,
            num_years: req.data_period.num_years,
        },
        completeness: req.completeness,
        is_us: req.is_us,
        surface_station: None,
        upper_air_station: None,
        status: "created".to_string(),
        steps_completed: vec![],
        errors: vec![],
        results: serde_json::Value::Object(serde_json::Map::new()),
    };

    save_project(&project)?;
    Ok(project)
}

pub fn load_project(project_id: &str) -> Result<Project, String> {
    let path = projects_dir().join(project_id).join("project.json");
    if !path.exists() {
        return Err(format!("Project {project_id} not found"));
    }
    let data = fs::read_to_string(&path).map_err(|e| format!("Read error: {e}"))?;
    serde_json::from_str(&data).map_err(|e| format!("Parse error: {e}"))
}

pub fn update_project(project: &mut Project, updates: serde_json::Value) -> Result<(), String> {
    if let serde_json::Value::Object(map) = updates {
        for (key, val) in map {
            match key.as_str() {
                "status" => {
                    if let Some(s) = val.as_str() {
                        project.status = s.to_string();
                    }
                }
                "surface_station" => project.surface_station = Some(val),
                "upper_air_station" => project.upper_air_station = Some(val),
                "results" => project.results = val,
                "errors" => {
                    if let Ok(e) = serde_json::from_value(val) {
                        project.errors = e;
                    }
                }
                _ => {}
            }
        }
    }
    save_project(project)
}

pub fn save_project(project: &Project) -> Result<(), String> {
    let path = Path::new(&project.dir).join("project.json");
    let data = serde_json::to_string_pretty(project).map_err(|e| format!("Serialize: {e}"))?;
    fs::write(&path, data).map_err(|e| format!("Write error: {e}"))
}
